server/transaction/chunk.go

```diff
@@ -0,0 +1,216 @@
+// anything for zeroth chunk. The chunk number is appended at the end in the format "__<chunk number>".
+type ChunkTx struct {
+	kv.Tx
+}
+
+func (tx *ChunkTx) executeChunks(data *internal.TableData, cb chunkCB) error {
+	if data.ActualUserPayloadSize() != int32(len(data.RawData)) {
+		return fmt.Errorf("unexpected size detected, actual user payload size '%d' is not matching the size set in data '%d'",
+			len(data.RawData), data.ActualUserPayloadSize())
+	}
+
+	originalDoc := data.RawData
+	marker := 0
+	chunk := int32(0)
+	for marker < len(originalDoc) {
+		end := marker + chunkSize
+		if end > len(originalDoc) {
+			end = len(originalDoc)
+		}
+
+		if err := cb(chunk, originalDoc[marker:end]); err != nil {
+			return err
+		}
+
+		marker += chunkSize
+		chunk++
+	}
+
+	return nil
+}
+
+func (tx *ChunkTx) setTotalChunks(data *internal.TableData) {
+	totalChunks := data.ActualUserPayloadSize() / int32(chunkSize)
+	if data.ActualUserPayloadSize()%int32(chunkSize) != 0 {
+		totalChunks += 1
+	}
+	data.TotalChunks = &totalChunks
+}
+
+func (tx *ChunkTx) Insert(ctx context.Context, key keys.Key, data *internal.TableData) error {
+	// first set total chunks
+	tx.setTotalChunks(data)
+
+	if *data.TotalChunks == 1 {
+		return tx.Tx.Insert(ctx, key.Table(), kv.BuildKey(key.IndexParts()...), data)
+	}
+
+	return tx.executeChunks(data, func(chunk int32, chunkData []byte) error {
+		if chunk == 0 {
+			// first part
+			zeroth := data.CloneForZerothChunk(chunkData)
+			if err := tx.Tx.Insert(ctx, key.Table(), kv.BuildKey(key.IndexParts()...), zeroth); err != nil {
+				return err
+			}
+		} else {
+			chunked := internal.NewChunkData(chunkData, chunk)
+			if err := tx.Tx.Replace(ctx, key.Table(), kv.BuildKey(append(key.IndexParts(), []any{chunkIdentifier, int(chunk)}...)...), chunked, false); err != nil {
+				return err
+			}
+		}
+
+		return nil
+	})
+}
+
+func (tx *ChunkTx) Replace(ctx context.Context, key keys.Key, data *internal.TableData, isUpdate bool) error {
+	// first set total chunks
+	tx.setTotalChunks(data)
+
+	if *data.TotalChunks == 1 {
+		return tx.Tx.Replace(ctx, key.Table(), kv.BuildKey(key.IndexParts()...), data, isUpdate)
+	}
+
+	// cleanup the existing range
+	if err := tx.Tx.Delete(ctx, key.Table(), kv.BuildKey(key.IndexParts()...)); err != nil {
+		return err
+	}
+
+	return tx.executeChunks(data, func(chunk int32, chunkData []byte) error {
+		var chunked *internal.TableData
+		var keyParts []any
+		if chunk == 0 {
+			// first part
+			chunked = data.CloneForZerothChunk(chunkData)
+			keyParts = key.IndexParts()
+		} else {
+			chunked = internal.NewChunkData(chunkData, chunk)
+			keyParts = append(key.IndexParts(), []any{chunkIdentifier, int(chunk)}...)
+		}
+
+		if err := tx.Tx.Replace(ctx, key.Table(), kv.BuildKey(keyParts...), chunked, isUpdate); err != nil {
+			return err
+		}
+
+		return nil
+	})
+}
+
+// Delete doesn't need any special handling as it is deleting on the key level not on chunk level.
```

Comment:
Just to confirm, the delete is going to delete a range based on the `PrimaryKey` so any chunk will be part of that range?

Reply:
that's right